File: 426. Convert Binary Search Tree to Sorted Doubly Linked List/treeToDoublyList.js
// Definition for a _Node.
class _Node {
  constructor(val, left, right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
};

/**
 * @param {_Node} root
 * @return {_Node}
 */
const treeToDoublyList = (root) => {
  if (!root) {
    return null;
  }
  let first = null;
  let last = null;
  const inOrderTraversal = (node) => {
    if (node) {
      inOrderTraversal(node.left);
      if (last) {
        last.right = node;
        node.left = last;
      } else {
        first = node;
      }
      last = node;
      inOrderTraversal(node.right);
    }
  };

  inOrderTraversal(root);
  last.right = first;
  first.left = last;

  return first;
};


const root = new _Node(4);
root.left = new _Node(2);
root.right = new _Node(5);

root.left.left = new _Node(1);
root.left.right = new _Node(3);

console.log('treeToDoublyList: ', treeToDoublyList(root));

File: 426. Convert Binary Search Tree to Sorted Doubly Linked List/treeToDoublyList.py
from typing import Optional

# Definition for a Node.
class Node:
  def __init__(self, val, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def treeToDoublyList(self, root: Optional[Node]) -> Optional[Node]:
    if not root:
      return None

    def inOrderTraversial(node):
      nonlocal first, last
      if node:
        inOrderTraversial(node.left)
        
        if last:
          last.right = node
          node.left = last
        else:
          first = node
        last = node
        
        inOrderTraversial(node.right)
    
    first, last = None, None  
    inOrderTraversial(root)
    first.left = last
    last.right = first

    return first
        
        
pTree = Node(4)
pTree.left = Node(2)
pTree.right = Node(5)

pTree.left.left = Node(1)
pTree.left.right = Node(3)

solution = Solution()
print('treeToDoublyList: ', solution.treeToDoublyList(pTree))
